File: ios/Chapbook/Sources/Chapbook/Transport.swift
import CChapbook
import Foundation

/// How a session reaches the network for catalogs and streamed pages.
///
/// The engine bundles a Rust transport (`ureq`) that is right for a
/// desktop process and wrong on iOS, where reaching the network outside
/// `URLSession` gives up background transfer, the system trust store, App
/// Transport Security, per-app VPN and the cellular-data toggle. So the
/// default is per-platform: `URLSession` on iOS, the bundled transport on
/// macOS. Most apps never name this type; an app with its own configured
/// session — an ephemeral one, a pinned one — passes it explicitly.
public struct HTTPTransport: Sendable {
    enum Kind: Sendable {
        case bundled
        case urlSession(URLSession)
    }

    let kind: Kind

    /// The engine's bundled Rust transport. The macOS default.
    public static let bundled = HTTPTransport(kind: .bundled)

    /// Fetch through the given session. Its configuration is honored
    /// wholesale — timeouts, caching, proxies, `waitsForConnectivity` —
    /// because the engine sends plain GETs and touches nothing else.
    public static func urlSession(_ session: URLSession) -> HTTPTransport {
        HTTPTransport(kind: .urlSession(session))
    }

    /// `URLSession.shared` everywhere the platform owns networking;
    /// [`bundled`](HTTPTransport.bundled) on macOS, where a plain process
    /// socket is the ordinary thing.
    public static var platformDefault: HTTPTransport {
        #if os(macOS)
            return .bundled
        #else
            return .urlSession(.shared)
        #endif
    }

    /// Install into a C config the caller still owns. On failure the
    /// engine has already run the finalizer — its ownership rule — so
    /// there is nothing to release here.
    func install(into config: OpaquePointer) throws {
        guard case .urlSession(let session) = kind else { return }
        let box = Unmanaged.passRetained(URLSessionTransport(session: session))
        try check(
            cb_config_set_http_transport(
                config, transportGet, transportDownload, transportFinalize,
                box.toOpaque()))
    }
}

/// The retained object behind the C `user` pointer. The engine promises
/// its finalizer runs exactly once — when the last session holding the
/// transport closes, or when installation fails — which is what makes
/// `passRetained`/`release` balance without this class counting anything.
final class URLSessionTransport: Sendable {
    let session: URLSession

    init(session: URLSession) {
        self.session = session
    }

    /// One blocking transfer — a session's GET, or the POST/PUT/DELETE a
    /// sync worker turns on; the method and body already ride on the
    /// request. The engine calls this on its loader or sync thread and
    /// expects the transfer settled on return; the semaphore bridges
    /// `URLSession`'s callback world to that contract. (A catalog *open*
    /// is therefore synchronous network — construct catalog sessions off
    /// the main actor.)
    func perform(_ request: URLRequest) -> Fetched {
        let slot = Slot<Fetched>()
        let done = DispatchSemaphore(value: 0)
        session.dataTask(with: request) { data, response, error in
            if let error {
                slot.value = .failure(error.localizedDescription)
            } else if let response = response as? HTTPURLResponse {
                slot.value = .response(response, data ?? Data())
            } else {
                slot.value = .failure("no HTTP response")
            }
            done.signal()
        }.resume()
        done.wait()
        return slot.value ?? .failure("the transfer never completed")
    }

    /// One blocking download, keeping the engine's promise: `dest` either
    /// ends up complete or is not created. `URLSession` lands the bytes
    /// in its own temp file; the hop to a `.part` sibling may cross
    /// volumes, but the final step is a same-directory rename, so no
    /// partial file ever carries the destination's name. A non-2xx status
    /// writes nothing — its body is not the book.
    func download(_ request: URLRequest, toPath dest: String) -> Fetched {
        let slot = Slot<Fetched>()
        let done = DispatchSemaphore(value: 0)
        session.downloadTask(with: request) { temp, response, error in
            defer { done.signal() }
            if let error {
                slot.value = .failure(error.localizedDescription)
                return
            }
            guard let response = response as? HTTPURLResponse else {
                slot.value = .failure("no HTTP response")
                return
            }
            guard (200..<300).contains(response.statusCode) else {
                slot.value = .response(response, Data())
                return
            }
            guard let temp else {
                slot.value = .failure("the platform delivered no file")
                return
            }
            // Inside the handler by necessity: the temp file dies when it
            // returns.
            let files = FileManager.default
            let part = dest + ".part"
            do {
                try? files.removeItem(atPath: part)
                try files.moveItem(atPath: temp.path, toPath: part)
                try files.moveItem(atPath: part, toPath: dest)
                slot.value = .response(response, Data())
            } catch {
                try? files.removeItem(atPath: part)
                slot.value = .failure("landing the download: \(error.localizedDescription)")
            }
        }.resume()
        done.wait()
        return slot.value ?? .failure("the transfer never completed")
    }

    enum Fetched {
        case response(HTTPURLResponse, Data)
        case failure(String)
    }

    /// Crosses the completion handler once: written before the semaphore
    /// signals, read after the wait, and the semaphore is the
    /// happens-before between them.
    private final class Slot<T>: @unchecked Sendable {
        var value: T?
    }
}

/// Rebuild the engine's request as a `URLRequest`, headers verbatim.
private func urlRequest(from request: cb_http_request) -> URLRequest? {
    guard let url = URL(string: String(cString: request.url)) else { return nil }
    var built = URLRequest(url: url)
    if let headers = request.headers {
        for header in UnsafeBufferPointer(start: headers, count: request.header_count) {
            built.setValue(
                String(cString: header.value), forHTTPHeaderField: String(cString: header.name))
        }
    }
    return built
}

/// Feed a fetch's outcome to the engine's response builder. 4xx and 5xx
/// go through as responses — a 401's body is the Authentication Document
/// the login flow needs — and only a transfer that produced nothing
/// reports as a failure. Every header crosses: a sync worker turns on
/// `ETag` and `Location`, and `URLSession` can enumerate them all as
/// cheaply as those two.
private func report(_ outcome: URLSessionTransport.Fetched, into response: OpaquePointer?) {
    switch outcome {
    case .failure(let message):
        _ = cb_http_response_fail(response, message)
    case .response(let http, let body):
        _ = cb_http_response_set_status(response, UInt16(clamping: http.statusCode))
        if let contentType = http.value(forHTTPHeaderField: "Content-Type") {
            _ = cb_http_response_set_content_type(response, contentType)
        }
        for (name, value) in http.allHeaderFields {
            guard let name = name as? String, let value = value as? String,
                name.caseInsensitiveCompare("Content-Type") != .orderedSame
            else { continue }
            _ = cb_http_response_add_header(response, name, value)
        }
        body.withUnsafeBytes { buffer in
            _ = cb_http_response_append_body(
                response, buffer.bindMemory(to: UInt8.self).baseAddress, buffer.count)
        }
    }
}

// The C entry points. Plain functions, not closures, so they carry no
// context — everything they need rides in `user`. Internal rather than
// private because `cb_sync_open` takes the same `get` and `finalize`,
// and `cb_catalog_open` those plus `download`.

func transportGet(
    request: UnsafePointer<cb_http_request>?,
    response: OpaquePointer?,
    user: UnsafeMutableRawPointer?
) {
    guard let request = request?.pointee, let user else { return }
    guard let built = urlRequest(from: request) else {
        _ = cb_http_response_fail(response, "the request URL did not parse")
        return
    }
    let transport = Unmanaged<URLSessionTransport>.fromOpaque(user).takeUnretainedValue()
    report(transport.perform(built), into: response)
}

/// The write half — `cb_http_send_fn`. Only a sync worker installs it:
/// reconciling marks means POST, PUT and DELETE against a Web Annotation
/// container, and a position PUT against a progression service.
func transportSend(
    method: UnsafePointer<CChar>?,
    request: UnsafePointer<cb_http_request>?,
    body: UnsafePointer<UInt8>?,
    bodyLen: Int,
    response: OpaquePointer?,
    user: UnsafeMutableRawPointer?
) {
    guard let method, let request = request?.pointee, let user else { return }
    guard var built = urlRequest(from: request) else {
        _ = cb_http_response_fail(response, "the request URL did not parse")
        return
    }
    built.httpMethod = String(cString: method)
    if let body, bodyLen > 0 {
        // Copied here by necessity: the engine's bytes die when this
        // callback returns, and the data task outlives the call frame.
        built.httpBody = Data(bytes: body, count: bodyLen)
    }
    let transport = Unmanaged<URLSessionTransport>.fromOpaque(user).takeUnretainedValue()
    report(transport.perform(built), into: response)
}

func transportDownload(
    request: UnsafePointer<cb_http_request>?,
    dest: UnsafePointer<CChar>?,
    response: OpaquePointer?,
    user: UnsafeMutableRawPointer?
) {
    guard let request = request?.pointee, let dest, let user else { return }
    guard let built = urlRequest(from: request) else {
        _ = cb_http_response_fail(response, "the request URL did not parse")
        return
    }
    let transport = Unmanaged<URLSessionTransport>.fromOpaque(user).takeUnretainedValue()
    report(transport.download(built, toPath: String(cString: dest)), into: response)
}

func transportFinalize(user: UnsafeMutableRawPointer?) {
    guard let user else { return }
    Unmanaged<URLSessionTransport>.fromOpaque(user).release()
}

File: windows/Chapbook/Transport.cs
using System.Runtime.InteropServices;
using System.Text;

namespace Chapbook;

/// <summary>One request the engine wants performed.</summary>
/// <param name="Url">
/// Opaque, and it may embed a per-user key: do not log it, and key any
/// credential by origin rather than by URL.
/// </param>
/// <param name="Headers">
/// What the engine wants sent, in order. Duplicates are meaningful.
/// </param>
public readonly record struct HttpRequest(
    string Url, IReadOnlyList<KeyValuePair<string, string>> Headers);

/// <summary>
/// The response being built, handed to a transport for the duration of one
/// callback.
/// </summary>
/// <remarks>
/// It is the one exception to "a transport must not call back into the
/// ABI": these builders, on this object, from inside the callback that was
/// given it. Using it after the callback returns is use-after-free.
/// </remarks>
public readonly ref struct HttpResponseBuilder
{
    private readonly nint _handle;

    internal HttpResponseBuilder(nint handle) => _handle = handle;

    /// <summary>
    /// The HTTP status, including 4xx and 5xx — those are responses, not
    /// failures, and the engine has its own flow for them.
    /// </summary>
    public void SetStatus(ushort status) =>
        ChapbookException.Check(
            Interop.cb_http_response_set_status(_handle, status), nameof(SetStatus));

    /// <summary>
    /// The <c>Content-Type</c> verbatim, parameters and all. Skip it when
    /// the server sent none; this value is authoritative over whatever the
    /// request asked for.
    /// </summary>
    public void SetContentType(string contentType) =>
        ChapbookException.Check(
            Interop.cb_http_response_set_content_type(_handle, contentType),
            nameof(SetContentType));

    /// <summary>Report one response header, name and value as received.</summary>
    /// <remarks>
    /// For the write half this is not optional. A Web Annotation container
    /// carries its whole concurrency story in <c>ETag</c> and says where it
    /// put a new mark in <c>Location</c>; a transport that discards them
    /// makes safe concurrent editing impossible, and the failure looks like
    /// sync quietly forgetting marks rather than like a missing header.
    /// </remarks>
    public void AddHeader(string name, string value) =>
        ChapbookException.Check(
            Interop.cb_http_response_add_header(_handle, name, value), nameof(AddHeader));

    /// <summary>Append body bytes. Call as many times as the body arrives in.</summary>
    public void AppendBody(ReadOnlySpan<byte> bytes)
    {
        if (bytes.IsEmpty)
        {
            return;
        }
        unsafe
        {
            fixed (byte* p = bytes)
            {
                ChapbookException.Check(
                    Interop.cb_http_response_append_body(_handle, (nint)p, (nuint)bytes.Length),
                    nameof(AppendBody));
            }
        }
    }

    /// <summary>
    /// The request never produced a response — DNS, the connection, a
    /// timeout. Not for 4xx or 5xx, which are responses.
    /// </summary>
    public void Fail(string message) =>
        ChapbookException.Check(
            Interop.cb_http_response_fail(_handle, message), nameof(Fail));
}

/// <summary>
/// The engine's networking, supplied by the host.
/// </summary>
/// <remarks>
/// <para>
/// A bundled networking stack is precisely the one a host cannot
/// substitute, so the engine opens no sockets of its own unless asked to.
/// On .NET the reason to substitute is concrete: an <see cref="HttpClient"/>
/// the app configured carries its proxy, its trust decisions, its
/// authentication handler and its timeouts, and a stack inside the engine
/// would carry none of them.
/// </para>
/// <para>
/// <b>The contract is unforgiving in three places.</b> A callback may fire
/// on any thread, including the engine's loader thread, and must
/// <i>block</i> until the transfer settles — this is not a place to return
/// a <see cref="Task"/>. It must not call back into this library except
/// through the <see cref="HttpResponseBuilder"/> it was handed. And it must
/// not retry: authentication retry is the engine's own flow one level up,
/// and a transport that retries turns one 401 into several.
/// </para>
/// </remarks>
public abstract class HttpTransport
{
    /// <summary>Perform a GET and report what came back.</summary>
    public abstract void Get(in HttpRequest request, HttpResponseBuilder response);

    /// <summary>
    /// Perform a POST, PUT or DELETE — the write half a sync transport must
    /// have, because reconciling marks means all three against a Web
    /// Annotation container and a position PUT against a progression
    /// service.
    /// </summary>
    /// <param name="method">
    /// An uppercase token, and only ever <c>POST</c>, <c>PUT</c> or
    /// <c>DELETE</c>.
    /// </param>
    /// <param name="body">The bytes to send; empty for a DELETE.</param>
    /// <remarks>
    /// Everything <see cref="Get"/> promises applies, plus reporting the
    /// response headers through
    /// <see cref="HttpResponseBuilder.AddHeader"/>.
    /// </remarks>
    public abstract void Send(
        string method, in HttpRequest request, ReadOnlySpan<byte> body,
        HttpResponseBuilder response);

    /// <summary>
    /// The engine has dropped this transport and will never call it again.
    /// </summary>
    /// <remarks>
    /// Called exactly once, from the ABI's finalizer: when the last session
    /// holding the transport closes, when a sync worker closes, or when a
    /// configuration is disposed without ever being opened. A host that
    /// created something for the engine's sake — a dedicated
    /// <see cref="HttpClient"/>, a handler with its own connection pool —
    /// releases it here, which is the only moment it is certain no callback
    /// is still running.
    /// </remarks>
    protected internal virtual void OnReleased()
    {
    }
}

/// <summary>
/// A transport over an <see cref="HttpClient"/> the host owns.
/// </summary>
/// <remarks>
/// The client is not disposed here: it belongs to whoever passed it, and
/// an <see cref="HttpClient"/> is meant to be long-lived and shared. A
/// service behind authentication wants one with a handler that attaches
/// it — no credential crosses this boundary, by design.
/// </remarks>
public class HttpClientTransport(HttpClient client) : HttpTransport
{
    private readonly HttpClient _client = client;

    /// <summary>A transport over a client this instance then owns for its lifetime.</summary>
    public HttpClientTransport() : this(new HttpClient())
    {
    }

    public override void Get(in HttpRequest request, HttpResponseBuilder response)
    {
        using var message = new HttpRequestMessage(HttpMethod.Get, request.Url);
        Apply(message, request.Headers);
        Perform(message, response);
    }

    public override void Send(
        string method, in HttpRequest request, ReadOnlySpan<byte> body,
        HttpResponseBuilder response)
    {
        using var message = new HttpRequestMessage(new HttpMethod(method), request.Url);
        Apply(message, request.Headers);
        if (!body.IsEmpty)
        {
            message.Content = new ByteArrayContent(body.ToArray());
        }
        Perform(message, response);
    }

    private static void Apply(
        HttpRequestMessage message, IReadOnlyList<KeyValuePair<string, string>> headers)
    {
        foreach ((string name, string value) in headers)
        {
            // Content headers are refused on the request collection, so
            // they go where they belong instead of being dropped.
            if (!message.Headers.TryAddWithoutValidation(name, value))
            {
                message.Content ??= new ByteArrayContent([]);
                message.Content.Headers.TryAddWithoutValidation(name, value);
            }
        }
    }

    private void Perform(HttpRequestMessage message, HttpResponseBuilder response)
    {
        try
        {
            // Blocking on purpose: the contract says this call must not
            // return until the transfer settles. It runs on an engine
            // thread with no synchronization context, so there is nothing
            // here to deadlock against.
            using HttpResponseMessage result =
                _client.Send(message, HttpCompletionOption.ResponseHeadersRead);

            response.SetStatus((ushort)result.StatusCode);
            if (result.Content.Headers.ContentType is { } contentType)
            {
                response.SetContentType(contentType.ToString());
            }
            foreach ((string name, IEnumerable<string> values) in
                     result.Headers.Concat(result.Content.Headers))
            {
                foreach (string value in values)
                {
                    response.AddHeader(name, value);
                }
            }

            using Stream stream = result.Content.ReadAsStream();
            byte[] buffer = new byte[64 * 1024];
            int read;
            while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
            {
                response.AppendBody(buffer.AsSpan(0, read));
            }
        }
        catch (Exception e) when (e is HttpRequestException or IOException
                                    or TaskCanceledException or InvalidOperationException)
        {
            // No response at all — DNS, the connection, a timeout. A 4xx
            // is not this; it went through `SetStatus` above.
            response.Fail(e.Message);
        }
    }
}

/// <summary>
/// The plumbing that puts a <see cref="HttpTransport"/> behind the ABI's
/// three function pointers.
/// </summary>
/// <remarks>
/// The transport instance is pinned by a <see cref="GCHandle"/> and that
/// handle is the ABI's <c>user</c> pointer. The engine calls
/// <c>finalize</c> exactly once when it drops the transport — a
/// configuration freed unopened, or the last session holding it closed —
/// and that is where the handle is released. This is the whole reason the
/// ABI has a finalizer: a host hands over a reference-counted object
/// without having to guess at the engine's lifetimes.
/// </remarks>
internal static unsafe class TransportBridge
{
    internal static nint Pin(HttpTransport transport) =>
        GCHandle.ToIntPtr(GCHandle.Alloc(transport));

    internal static delegate* unmanaged<nint, nint, nint, void> Get => &OnGet;

    internal static delegate* unmanaged<nint, nint, nint, nuint, nint, nint, void> Send => &OnSend;

    internal static delegate* unmanaged<nint, void> Finalize => &OnFinalize;

    [UnmanagedCallersOnly]
    private static void OnGet(nint request, nint response, nint user)
    {
        // Nothing may unwind into Rust. A host's exception becomes a
        // reported failure, which is a thing the engine can act on, rather
        // than a torn process.
        var builder = new HttpResponseBuilder(response);
        try
        {
            if (GCHandle.FromIntPtr(user).Target is HttpTransport transport)
            {
                transport.Get(Read(request), builder);
            }
        }
        catch (Exception e)
        {
            Report(builder, e);
        }
    }

    [UnmanagedCallersOnly]
    private static void OnSend(
        nint method, nint request, nint body, nuint bodyLength, nint response, nint user)
    {
        var builder = new HttpResponseBuilder(response);
        try
        {
            if (GCHandle.FromIntPtr(user).Target is HttpTransport transport)
            {
                ReadOnlySpan<byte> bytes = body == 0
                    ? []
                    : new ReadOnlySpan<byte>((void*)body, checked((int)bodyLength));
                transport.Send(
                    Marshal.PtrToStringUTF8(method) ?? "GET", Read(request), bytes, builder);
            }
        }
        catch (Exception e)
        {
            Report(builder, e);
        }
    }

    [UnmanagedCallersOnly]
    private static void OnFinalize(nint user)
    {
        try
        {
            if (user != 0)
            {
                GCHandle handle = GCHandle.FromIntPtr(user);
                (handle.Target as HttpTransport)?.OnReleased();
                handle.Free();
            }
        }
        catch
        {
            // Deliberately swallowed; see above.
        }
    }

    private static void Report(HttpResponseBuilder builder, Exception e)
    {
        try
        {
            builder.Fail(e.Message);
        }
        catch
        {
            // The response may already have been failed or completed. There
            // is nowhere left to say anything.
        }
    }

    private static HttpRequest Read(nint request)
    {
        NativeHttpRequest native = *(NativeHttpRequest*)request;
        string url = Marshal.PtrToStringUTF8(native.Url) ?? string.Empty;
        var headers = new List<KeyValuePair<string, string>>((int)native.HeaderCount);
        for (nuint i = 0; i < native.HeaderCount; i++)
        {
            NativeHttpHeader header = ((NativeHttpHeader*)native.Headers)[i];
            headers.Add(new KeyValuePair<string, string>(
                Marshal.PtrToStringUTF8(header.Name) ?? string.Empty,
                Marshal.PtrToStringUTF8(header.Value) ?? string.Empty));
        }
        return new HttpRequest(url, headers);
    }
}

public sealed partial class SessionConfiguration
{
    /// <summary>
    /// Fetch through the host's networking instead of the bundled
    /// transport — for OPDS catalogues and streamed comic pages.
    /// </summary>
    /// <remarks>
    /// The transport is owned by the engine from this call onward and is
    /// released when the last session holding it closes, or when this
    /// configuration is disposed unopened.
    /// </remarks>
    public SessionConfiguration WithTransport(HttpTransport transport)
    {
        ArgumentNullException.ThrowIfNull(transport);
        nint user = TransportBridge.Pin(transport);
        unsafe
        {
            ChapbookException.Check(
                Interop.cb_config_set_http_transport(
                    Live(),
                    (nint)TransportBridge.Get,
                    (nint)TransportBridge.Finalize,
                    user),
                nameof(WithTransport));
        }
        return this;
    }
}

[StructLayout(LayoutKind.Sequential)]
internal struct NativeHttpHeader
{
    public nint Name;
    public nint Value;
}

[StructLayout(LayoutKind.Sequential)]
internal struct NativeHttpRequest
{
    public nint Url;
    public nint Headers;
    public nuint HeaderCount;
}
